File: Java/Algorithm/Algorithm I/206. Reverse Linked List.java
class Solution {

  public ListNode reverseList(ListNode head) {
    ListNode currentNode = head;
    ListNode previousNode = null;

    while (currentNode != null) {
      ListNode tempNode = currentNode.next;
      currentNode.next = previousNode;

      previousNode = currentNode;
      currentNode = tempNode;
    }

    return previousNode;
  }
}

File: Python/Algorithm/Algorithm I/206. Reverse Linked List.py
from typing import Optional


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """
        Reverse a singly-linked list in-place.

        Args:
            head (Optional[ListNode]): The head of the linked list.

        Returns:
            Optional[ListNode]: The head of the reversed linked list.

        Time Complexity: o(n) where n is the length of the linked list.
        Space Complexity: o(1) since we are not using any additional space.
        """
        previousNode = None
        currentNode = head

        while currentNode:
            nextNode = currentNode.next
            currentNode.next = previousNode

            previousNode = currentNode
            currentNode = nextNode

        return previousNode
